File: packages/swift/Sources/PRCProtocol/ServerAuth.swift
import Foundation
import PRCIdentity

/// Authentication of a device to the rendezvous server (spec section 11.1).
public enum ServerAuth {
    public static let context = "prc-server-auth-v1"
    public static let nonceBytes = 32
    public static let nonceTtlMs: Int64 = 60_000

    public static func signingInput(nonce: String, origin: String, deviceId: String) -> Data {
        Data("\(context)\n\(nonce)\n\(origin)\n\(deviceId)".utf8)
    }

    public static func sign(identity: any SigningIdentity, nonce: String, origin: String) throws -> String {
        Base64URL.encode(try identity.sign(signingInput(nonce: nonce, origin: origin, deviceId: identity.deviceId)))
    }

    public static func verify(publicKeyRaw: Data, nonce: String, origin: String, deviceId: String, signature: String) -> Bool {
        guard let sig = Base64URL.decode(signature) else { return false }
        return Verifier.verify(publicKeyRaw: publicKeyRaw, data: signingInput(nonce: nonce, origin: origin, deviceId: deviceId), signature: sig)
    }
}

File: packages/protocol/src/server-auth.ts
/** Authentication of a device to the rendezvous server (spec section 11.1). */
import { b64urlDecode, b64urlEncode, utf8Encode } from './encoding.ts';
import { sign, verify, type Identity } from './identity.ts';

export const SERVER_AUTH_CONTEXT = 'prc-server-auth-v1';
export const SERVER_NONCE_BYTES = 32;
export const SERVER_NONCE_TTL_MS = 60_000;

export function serverAuthSigningInput(nonce: string, origin: string, deviceId: string): Uint8Array {
  return utf8Encode(`${SERVER_AUTH_CONTEXT}\n${nonce}\n${origin}\n${deviceId}`);
}

export async function signServerAuth(identity: Identity, nonce: string, origin: string): Promise<string> {
  return b64urlEncode(await sign(identity.privateKey, serverAuthSigningInput(nonce, origin, identity.deviceId)));
}

export async function verifyServerAuth(publicKey: CryptoKey, nonce: string, origin: string, deviceId: string, sigB64: string): Promise<boolean> {
  let sig: Uint8Array;
  try {
    sig = b64urlDecode(sigB64);
  } catch {
    return false;
  }
  return verify(publicKey, serverAuthSigningInput(nonce, origin, deviceId), sig);
}
